Add a process for becoming a maintainer (#960)
open-telemetry/community: community-membership.md

### Requirements

> Unlike the roles outlined above, the Owners of a subproject are typically

Reviewer: nit: are we calling these people Owners or Maintainers? (I prefer Maintainers since we've already used that term above) it seems (?) like we're using those terms interchangeably here.

Author: I took the wording from the "requirements" section above. I'll change them both.

limited to a relatively small group of decision makers and updated as fits
the needs of the subproject.

> The following apply to the subproject for which one would be an owner.

Reviewer: s/owner/maintainer/ (per the above)?

- Work with other maintainers to maintain the project's overall health and
  success holistically.

### Becoming a Maintainer

Unless stated otherwise in a SIG charter ratified by the Technical Committee,

> a new maintainer is elected by unanimous vote of the existing maintainers

Reviewer 2: I don't have a strong opinion about the "unanimous vote" part, but for the reference the Jaeger's process is a bit more relaxed:

> Two other maintainers need to second the nomination. If no one objects in 5 working days (U.S.), the nomination is accepted. If anyone objects or wants more information, the maintainers discuss and usually come to a consensus (within the 5 working days). If issues can't be resolved, there's a simple majority vote among current maintainers.

Author: Our project is broken into many small groups who each maintain only one portion with only a few maintainers. Unanimous consent of one SIG represents only a small portion of the overall maintainers and shouldn't be too high of a bar I think.

Reviewer 3: I like the lazy consensus part of @yurishkuro 's quote. Active consensus sometimes hard to get.

Author: For most things I'd agree but it seems like adding a maintainer is about the most impactful thing you can do. If a maintainer can't be bothered to consent to a new maintainer either the new maintainer isn't agreed on or the existing maintainer isn't doing their job imo

Reviewer 4: > or the existing maintainer isn't doing their job

This might indeed happen: maintainers who have moved on might not have yet turned themselves into emeritus and might be hard to reach. Same for maintainers on longer breaks.

Author: So how about if there is a unanimous vote then the vote passes, otherwise if there is no objection within 5 days the vote passes?

Author: I updated the PR to reflect this lazy consensus. PTAL

of the SIG. After a successful vote, a pull request should be made adding
the new maintainer to the list of existing maintainers. Before the pull request
is merged, all existing maintainers and the newly elected maintainer should
approve the pull request to show their support for the public record. The
nominee is considered a maintainer after the pull request is merged.

> 

Reviewer 2: ```suggestion

#### Self-nomination is possible and encouraged

```

> If you feel like you meet the requirements above and are willing to take on the

Reviewer 2: > requirements above

I don't think "above" actually lists any requirements to _becoming a maintainer_, it only lists responsibilities of _being_ a maintainer. In Jaeger we had [specific requirements][1] that must be met, where the actual required stats were less important than the fact that they are written down (for a fair process).

Do we want to propose an org-level maintainer stats requirements so that SIGs won't need to define their own (but still have the option to do so)?

[1]: https://github.com/jaegertracing/jaeger/blob/main/GOVERNANCE.md#becoming-a-maintainer

Author: https://github.com/open-telemetry/community/blob/main/community-membership.md#requirements-3

Reviewer 2: It's not qualitative, and in fact the only line related to direct code contributions (which I would consider a must for all non-spec repos) is somewhat weak:

> Directly contributed to the subproject through implementation and / or review

Author: i think the requirements of the position should be a separate PR from the process itself

Reviewer 2: +1

additional responsibilities and privileges of being a maintainer, it is
recommended that you approach an existing maintainer about sponsoring your bid
to become a maintainer. After you and your sponsor have discussed the role

> and its additional requirements and responsibilities, they may approach the other

Reviewer 4: Should we list the common expectations connected to this role? Like:

- performing releases
- answering users questions
- fixing bugs across the entire code base
- mentoring other prospective maintainers

Author: Isn't that listed above in requrements/responsibilities/privileges sections?

Reviewer 4: Sorry about that, it indeed is!

maintainers about a vote to confirm you as a new maintainer. If the maintainer
does not believe you are ready for the role, or the subproject is not in need

> of additional maintainers, they may suggest an alternate role or areas you can
> improve in order to improve your chances to become a maintainer in the future.

Reviewer 2: ```suggestion
of additional maintainers, they may suggest an alternate role or growth areas 
to improve your chances to become a maintainer in the future.
```

a new maintainer is elected by unanimous vote of the existing maintainers
of the SIG. After a successful vote, a pull request should be made adding
the new maintainer to the list of existing maintainers. Before the pull request
is merged, all existing maintainers and the newly elected maintainer should
approve the pull request to show their support for the public record. The
nominee is considered a maintainer after the pull request is merged.

#### Self-nomination is encouraged

If you feel like you meet the requirements above and are willing to take on the
additional responsibilities and privileges of being a maintainer, it is
recommended that you approach an existing maintainer about sponsoring your bid
to become a maintainer. After you and your sponsor have discussed the role
and its additional requirements and responsibilities, they may approach the other
maintainers about a vote to confirm you as a new maintainer. If the maintainer
does not believe you are ready for the role, or the subproject is not in need
of additional maintainers, they may suggest an alternate role or growth areas
in order to improve your chances to become a maintainer in the future.

> ### Resolving technical conflicts within a SIG

Reviewer: Given the requirement of unanimity above, I'd suggest that we augment this section to include "non-technical conflicts" (i.e., have some escalation process for removing someone as a maintainer in a SIG).

... Or, alternatively, reduce the requirement for adding a maintainer to be less than a unanimous vote of existing maintainers?

Author: It seems others feel the same way about the unanimous vote so I loosened the restrictions. PTAL.

- Work with other maintainers to maintain the project's overall health and
  success holistically.

### Becoming a Maintainer

Unless stated otherwise in a SIG charter ratified by the Technical Committee,
a new maintainer is elected by vote of the existing maintainers of the SIG.
The vote is officially started when a pull request to add a new maintainer
is opened, and ends when the pull request is merged. The pull request may be
merged when the following conditions are met:

- The person being nominated has accepted the nomination by approving the pull request

> - All maintainers have approved the pull request OR a majority of maintainers

Reviewer 5: I am not sure I understand the "OR" here. It seems like the right-side of the OR is a sufficient condition and the left-side is superfluous.

Author: if all maintainers approve you don't need to wait the full 5 days. should we just wait 5 days no matter what?

Reviewer 5: Ah, I see, there is a difference then. I think what you have is fine.

- Work with other maintainers to maintain the project's overall health and
  success holistically.

### Becoming a Maintainer

Unless stated otherwise in a SIG charter ratified by the Technical Committee,
a new maintainer is elected by vote of the existing maintainers of the SIG.
The vote is officially started when a pull request to add a new maintainer
is opened, and ends when the pull request is merged. The pull request may be
merged when the following conditions are met:

- The person being nominated has accepted the nomination by approving the pull request
- All maintainers have approved the pull request OR a majority of maintainers

>   have approved the pull request and no maintainer has objected by requesting changes on the pull request

Author: You're right. Added 3e035710659d76dab81631a3bfcd8b3716fa5316
